# 1.1 Setup
## Importación de librerías

In [ ]:
import pandas as pd
import numpy as np
import csv


Carga

In [ ]:
df = pd.read_csv(
    "../data/ventas_techventas.csv",
    encoding="latin1"
)

df.head()

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
0,1001,2024-01-05,C001,TechCorp SA,Norte,Laptop Pro 15,Electronica,2.0,1200.0,0.10,Ana L�ï¿
1,1002,2024-01-07,C002,Innovatek,Sur,Mouse Inalambrico,Perifericos,15.0,25.0,0.00,Carlos Ruiz
2,"1003,2024-01-10,C003,DataSolutions,Centro,Moni...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,1004,2024-01-12,C001,TechCorp SA,Norte,Teclado Mecanico,Perifericos,10.0,85.0,0.10,Luis Mora
4,1005,2024-01-15,C004,CloudNet,Este,Laptop Pro 15,Electronica,3.0,1200.0,0.15,Carlos Ruiz


## Detección de nulos y negativos

In [ ]:
df[
    df.isna().any(axis=1) |
    (df["cantidad"] < 0)
]

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
42,1043,2024-05-01,C010,DeltaOps,Norte,NaN,Electronica,2,350.0,0.05,Maria Torres
47,1048,2024-05-15,C007,AlphaTech,Sur,Laptop Pro 15,Electronica,-1,1200.0,0.00,Ana L�ï¿


## Tratamiento de negativos y nulos, y normalización de fechas

In [ ]:
# Recargar el archivo ignorando comillas separadoras
df = pd.read_csv(
    "ventas_techventas.csv",
    quoting=csv.QUOTE_NONE,
    encoding="latin1"
)

df.head()

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
0,1001,2024-01-05,C001,TechCorp SA,Norte,Laptop Pro 15,Electronica,2,1200.0,0.10,Ana L�ï¿
1,1002,2024-01-07,C002,Innovatek,Sur,Mouse Inalambrico,Perifericos,15,25.0,0.00,Carlos Ruiz
2,"""1003",2024-01-10,C003,DataSolutions,Centro,"Monitor 27""""",Electronica,5,350.0,0.05,"Ana L�ï¿"""
3,1004,2024-01-12,C001,TechCorp SA,Norte,Teclado Mecanico,Perifericos,10,85.0,0.10,Luis Mora
4,1005,2024-01-15,C004,CloudNet,Este,Laptop Pro 15,Electronica,3,1200.0,0.15,Carlos Ruiz


## Detección de nulos y negativos de la nueva carga

In [ ]:
df[
    df.isna().any(axis=1) |
    (df["cantidad"] < 0)
]

,order_id,fecha,cliente_id,cliente_nombre,region,producto,categoria,cantidad,precio_unitario,descuento,vendedor
42,1043,2024-05-01,C010,DeltaOps,Norte,NaN,Electronica,2,350.0,0.05,Maria Torres
47,1048,2024-05-15,C007,AlphaTech,Sur,Laptop Pro 15,Electronica,-1,1200.0,0.00,Ana L�ï¿


In [ ]:
print("tamaño del dataset original", df.size)
''' Para producto nulo en el registro 42 se realiza:
Rellenar el nulo basado en la coincidencia de precio y categoría '''
df.loc[(df['producto'].isna()) & (df['precio_unitario'] == 350.0), 'producto'] = 'Monitor 27'

# Para cantidades negativas, se elimina el registro
df = df[df["cantidad"] > 0]

# Normalización de fechas
df["fecha"] = pd.to_datetime(
    df["fecha"],
    errors="coerce"
)
df = df[df["fecha"].notna()]
print("tamaño del dataset limpio", df.size)

tamaño del dataset original 649
tamaño del dataset limpio 649


# 1.2 Álgebra Lineal Aplicada
## Arrays

In [ ]:
cantidades = df["cantidad"].to_numpy()
precios = df["precio_unitario"].to_numpy()

## Cálculo de Ingresos Brutos

In [ ]:
# con operador @
ingresos = cantidades @ precios
print("ingresos totales: ",ingresos)

ingresos_por_venta = cantidades * precios
print(ingresos_por_venta)

ingresos totales:  89405.0
[2400.  375. 1750.  850. 3600. 1900.  700.  960. 1140. 1200. 1400.  750.
  425.  360. 2375. 2400. 2100. 1710.  300. 2400.  480. 2850.  680. 1050.
 4800.  500. 1425.  720. 1200.  850. 1050. 2090. 6000.  625.  240. 1400.
 1020. 1710.  700. 3600.  375.  600.  700. 3600. 1900.  510.  450.  480.
 2375. 1050. 2400.  500. 3325. 1750. 2400.  840.  765. 1900. 1400.]


## 1.3 Estadísticas con NumPy

In [ ]:
precios = df["precio_unitario"].dropna().to_numpy()

print("Media:", np.mean(precios))
print("Mediana:", np.median(precios))
print("Std:", np.std(precios))
print("P25:", np.percentile(precios,25))
print("P75:", np.percentile(precios,75))

Media: 364.3220338983051
Mediana: 120.0
Std: 435.6560002243243
P25: 90.0
P75: 350.0


## Interpretación de las estadísticas descriptivas

Se calcularon la media, mediana, desviación estándar y percentiles del campo **precio_unitario** utilizando NumPy.

### Resultados obtenidos

| Métrica | Valor |
|----------|----------:|
| Media | 364.32 |
| Mediana | 120.00 |
| Desviación estándar | 435.66 |
| Percentil 25 | 90.00 |
| Percentil 75 | 350.00 |

### Análisis

La **media** de los precios es de **364.32**, mientras que la **mediana** es de **120.00**. La diferencia significativa entre ambos valores indica que la distribución de precios no es simétrica y se encuentra sesgada hacia la derecha. Esto sugiere la existencia de productos de alto valor que elevan considerablemente el promedio general del catálogo.

La **desviación estándar** de **435.66** es incluso superior a la mediana, lo que evidencia una alta dispersión en los precios. En otras palabras, el catálogo contiene productos con rangos de precio muy diferentes, desde artículos relativamente económicos hasta productos considerablemente más costosos.

El **percentil 25 (P25)** indica que el 25% de los productos tienen precios iguales o inferiores a **90.00**, mientras que el **percentil 75 (P75)** muestra que el 75% de los productos tienen precios iguales o inferiores a **350.00**. Esto significa que el 50% central de los precios se concentra entre **90.00 y 350.00**.

### Conclusiones sobre el catálogo

Los resultados sugieren que TechVentas S.A.S. maneja un catálogo diversificado con productos de distintos segmentos de precio. Aunque la mayoría de los artículos se concentra en rangos moderados (entre 90 y 350 unidades monetarias), existen algunos productos de alto valor que incrementan significativamente el promedio general.

Esta situación es común en empresas tecnológicas donde conviven accesorios de bajo costo, como periféricos o componentes básicos, con equipos de mayor valor, como computadores, servidores o dispositivos especializados. Por ello, la mediana representa mejor el precio típico del catálogo que la media, ya que es menos sensible a valores extremos.